operator/pkg/ingress/secret.go

```diff
@@ -0,0 +1,363 @@
+		sm.queue.AddRateLimited(event)
+	} else {
+		log.Errorf("Failed to process Ingress event, skipping: %+v", event)
+		sm.queue.Forget(event)
+	}
+	return true
+}
+
+func (sm *syncSecretManager) handleEvent(event interface{}) interface{} {
+	var err error
+	switch ev := event.(type) {
+	case secretAddedEvent:
+		err = sm.handleSecretAddedEvent(ev)
+	case secretUpdatedEvent:
+		err = sm.handleSecretUpdatedEvent(ev)
+	case secretDeletedEvent:
+		err = sm.handleSecretDeletedEvent(ev)
+	case ingressAddedEvent:
+		err = sm.handleIngressAddedEvent(ev)
+	case ingressUpdatedEvent:
+		err = sm.handleIngressUpdatedEvent(ev)
+	case ingressDeletedEvent:
+		err = sm.handleIngressDeletedEvent(ev)
+	default:
+		err = fmt.Errorf("received an unknown event: %t", ev)
+	}
+	return err
+}
+
+func (sm *syncSecretManager) handleSecretAddedEvent(ev secretAddedEvent) error {
+	if ev.secret.GetNamespace() == sm.namespace {
+		return nil
+	}
+	return sm.syncSecret(ev.secret)
+}
+
+func (sm *syncSecretManager) handleSecretUpdatedEvent(ev secretUpdatedEvent) error {
+	if ev.oldSecret.GetNamespace() == sm.namespace || ev.newSecret.GetNamespace() == sm.namespace {
+		return nil
+	}
+	return sm.syncSecret(ev.newSecret)
+}
+
+func (sm *syncSecretManager) handleSecretDeletedEvent(ev secretDeletedEvent) error {
+	if ev.secret.GetNamespace() == sm.namespace {
+		return nil
+	}
+	return sm.deleteSecret(ev.secret)
+}
+
+func (sm *syncSecretManager) handleIngressAddedEvent(ev ingressAddedEvent) error {
+	return sm.handleIngressUpsertedEvent(ev.ingress)
+}
+
+func (sm *syncSecretManager) handleIngressUpdatedEvent(ev ingressUpdatedEvent) error {
+	return sm.handleIngressUpsertedEvent(ev.newIngress)
+}
+
+// handleIngressDeletedEvent is doing nothing right now as the underlying secrets could be
+// referenced in other Ingress resources, which is common use case for wildcard secrets (e.g
+// *.foo.com)
+func (sm *syncSecretManager) handleIngressDeletedEvent(ev ingressDeletedEvent) error {
+	return nil
+}
+
+func (sm *syncSecretManager) handleIngressUpsertedEvent(ingress *slim_networkingv1.Ingress) error {
+	if !tlsEnabled(ingress) {
+		return nil
+	}
+
+	for _, tls := range ingress.Spec.TLS {
+		// check if the secret is available
+		key := getSecretKey(ingress.GetNamespace(), tls.SecretName)
+		secret, exists, err := sm.getByKey(key)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return fmt.Errorf("secert does not exist: %s", key)
+		}
+
+		sm.lock.Lock()
+		sm.watchedSecretMap[key] = getSyncedSecretKey(sm.namespace, secret.GetNamespace(), secret.GetName())
+		sm.lock.Unlock()
+
+		// proceed to sync secret
+		err = sm.syncSecret(secret)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// syncSecret performs an upsert to make sure that secret in secret namespace is in synced with original secret
+func (sm *syncSecretManager) syncSecret(original *slim_corev1.Secret) error {
+	sm.lock.RLock()
+	defer sm.lock.RUnlock()
```

Comment:
Do we need to do a RLock until the function is returned?

Reply:
good point :100: